### PR TITLE
TBOX 61 Added email notifications

### PR DESCRIPTION
**Pull request by tristan-koopman**
TBOX 61

Created a new emails module under notifications similar to slack. Added TLS and SSL support. SSL is default.

Added new monitor function under utils.operation.monitor and a test for it

Welcome comments on the script, structure, how to improve, etc.

Looking for any feedback on the tests. Testing a mock smtp server is new to me.

Example of an email sent using the toolbox send_email function. (Using SSL and gmail smtp server)
Screen Shot 2021-10-05 at 3 04 16 PM

### CODE REVIEW

File: tamr_toolbox/notifications/email_info.py
Comment (along528) [inline tamr_toolbox/notifications/email_info.py]: The pattern used elsewhere in the toolbox is to use kwargs for each parameter expected in the config. This provides more flexibility with using the script and is easier to document. You can still pass the config object via unpacking:
```
def my_func(*, param1, param2):
  ...do a thing...

config = { "param1": "val1", "param2": "val2"}
my_func(**config)
```

See for example [tbox.utils.client.create](https://github.com/Datatamer/tamr-toolbox/blob/develop/tamr_toolbox/utils/client.py#L42) and it[ in use](https://github.com/Datatamer/tamr-toolbox/blob/develop/examples/scripts/utils/upstream/get_upstream_projects.py#L24)

File: tamr_toolbox/notifications/emails.py
Comment (along528) [inline tamr_toolbox/notifications/emails.py]: what's the reasoning for using a dataclass? Why not have each variable in dataclass as a parameter?

File: tamr_toolbox/notifications/email_info.py
Comment (along528) [inline tamr_toolbox/notifications/email_info.py]: Docstring needs a `Returns` block

File: tamr_toolbox/notifications/emails.py
Comment (along528) [inline tamr_toolbox/notifications/emails.py]: should this be an internal function? if so, should change to `_build_message`

File: tamr_toolbox/notifications/emails.py
Comment (along528) [inline tamr_toolbox/notifications/emails.py]: provide some more description

File: tamr_toolbox/notifications/emails.py
Comment (along528) [inline tamr_toolbox/notifications/emails.py:155]: This mostly overlaps with the slack `monitor_job` function. These should should refer to the same function. This new function should be internal in a new module `notifications.common`. See an example of this [here](https://github.com/Datatamer/tamr-toolbox/blob/develop/tamr_toolbox/data_io/common.py). The `slack.monitor_job` function will need to be updated as well, preserving backwards compatibility. 

As a suggestion, it should be possible to generalize this function such that the `_send_job_status_message` is the only piece that is unique. This could then be passed as a parameter. Any unique arguments to that function could be handled using [functools.partial](https://docs.python.org/3/library/functools.html#functools.partial) before passing as a parameter.

File: tamr_toolbox/notifications/email_info.py
Comment (along528) [inline tamr_toolbox/notifications/email_info.py]: If using a dataclass this should go in the [models/ directory](https://github.com/Datatamer/tamr-toolbox/tree/develop/tamr_toolbox/models)

**Review by along528 — CHANGES_REQUESTED** [review]
This is already in a great state! Only a few changes requested

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: I thought it would be annoying to pass the smtp configurations and addresses every time you called send email or monitor job. I can remove the dataclass if having the 5 individual parameters is more intuitive or better

Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: I made this change

File: tamr_toolbox/notifications/email_info.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/email_info.py]: email_info removed

File: tamr_toolbox/notifications/email_info.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/email_info.py]: email_info removed

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: Changed

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: Added more info

File: tamr_toolbox/notifications/email_info.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/email_info.py]: email_info removed

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py:155]: I generalized the function in notifications.common as suggested. Ended up using kwargs to account for the unique arguments. Let me know if functools.partial is better

File: examples/snippets/notifications/enable_email_notifications.py
Comment (skalish) [inline examples/snippets/notifications/enable_email_notifications.py:28]: For this example, it's worth calling out each argument as you are doing, but I want to highlight that because of the way `config` is constructed (matching key names to kwargs) you can do this concisely with
```suggestion
tbox.notifications.emails.send_email(
    message="This is a test message.",
    subject_line="Subject",
    **config,
)
```
If you had a more general config object, my recommendation would be to have your current one as a dictionary under `"mail_config"`, and then pass `**config["mail_config"]`.

Comment (tristan-koopman) [reply examples/snippets/notifications/enable_email_notifications.py:28]: That is very smart! Will use in the future

File: tamr_toolbox/notifications/emails.py
Comment (along528) [inline tamr_toolbox/notifications/emails.py]: You are no longer using a contextmanager via `with` as you were before. Why not?

File: tamr_toolbox/utils/operation.py
Comment (along528) [inline tamr_toolbox/utils/operation.py]: What happens for default timeout_seconds = None? Probably should use a default numeric value instead, e.g. 60 seconds

File: tamr_toolbox/utils/operation.py
Comment (along528) [inline tamr_toolbox/utils/operation.py:131]: Please add an example to the documentation for using this new function

**Review by along528 — COMMENTED** [review]
Looking good, just a couple of small comments

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: This was changed to accommodate the TLS code. Is there a simple way to use `with` when you have two different functions with different parameters? I tried to reduce the number of duplicate lines of code by removing `with`. 

Alternative:
```
if use_tls:
   with smtplib.SMTP as server:
   # repetitive code here
else:
   with smtplin.SMTP_SSL as server:
   # repetitive code also here
```

Comment (skalish) [reply tamr_toolbox/notifications/emails.py]: You could try:
```python
with smtplib.SMTP if use_tls else smtplib.SMTP_SSL as server:
   # repetitive code here
```

File: tamr_toolbox/utils/operation.py
Comment (tristan-koopman) [reply tamr_toolbox/utils/operation.py]: The operation would hang "forever". I updated the default to be 300 seconds

File: tamr_toolbox/utils/operation.py
Comment (tristan-koopman) [reply tamr_toolbox/utils/operation.py:131]: Added example snippets for wait and monitor

File: examples/snippets/utils/operation/job_operations.py
Comment (along528) [inline examples/snippets/utils/operation/job_operations.py:11]: Add an extra comment or two on what monitor is doing exactly. Namely, returning when there is a state change

Also for wait

File: tamr_toolbox/utils/operation.py
Comment (along528) [inline tamr_toolbox/utils/operation.py]: With latest default, change type for `timeout_seconds`  to `int` instead of `Optional[int]`

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: Thanks!

File: examples/snippets/utils/operation/job_operations.py
Comment (tristan-koopman) [reply examples/snippets/utils/operation/job_operations.py:11]: Thanks. Done

File: tamr_toolbox/utils/operation.py
Comment (tristan-koopman) [reply tamr_toolbox/utils/operation.py]: Great catch. Done

File: examples/snippets/utils/operation/job_operations.py
Comment (skalish) [inline examples/snippets/utils/operation/job_operations.py]:
```suggestion
# OperationState.CANCELED, OperationState.SUCCEEDED, OperationState.FAILED
```
Typo

File: tamr_toolbox/notifications/common.py
Comment (skalish) [inline tamr_toolbox/notifications/common.py]:
```suggestion
    """Monitors a Tamr Operation and sends a message when the job status is updated
```
Fix formatting

File: tamr_toolbox/notifications/emails.py
Comment (skalish) [inline tamr_toolbox/notifications/emails.py]:
```suggestion
    """Checks operation state and if in `notify_states` sends the message.
```
Unnecessary newline

File: tamr_toolbox/notifications/emails.py
Comment (skalish) [inline tamr_toolbox/notifications/emails.py]:
```suggestion
    """Monitors a Tamr Operation and sends an email when the job status is updated
```

File: tamr_toolbox/notifications/emails.py
Comment (along528) [inline tamr_toolbox/notifications/emails.py]: Should `use_tls=True` be the default? Is there any additional configuration required to be able to use this?

File: tamr_toolbox/notifications/emails.py
Comment (along528) [inline tamr_toolbox/notifications/emails.py]: If I'm reading this correctly https://docs.python.org/3/library/smtplib.html#smtplib.SMTP_SSL then it is not necessary to call `server.starttls` when using `smtplib.SMTP_SSL`

File: tamr_toolbox/notifications/emails.py
Comment (along528) [inline tamr_toolbox/notifications/emails.py]: Add the optional variables `keyfile` and `certfile` with a default of `None` to be passed to the SMTP_SSL function. This way a user could specify these if they want to authenticate at the client-side and not just on the server side. See https://stackoverflow.com/questions/24250020/sending-email-without-keyfile-only-certfile-using-python-smtplib

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: SSL and TLS are two different cryptographic protocols. You can use one or the other. I have no problem with TLS being the default if you want. TLS is more secure. The only difference in terms of configuration is they use different ports. TLS is 587 and SSL is 465. All of the other configurations would be done on the SMTP server side.

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: That is correct. If you are using TLS, you don't use `smtplib.SMTP_SSL`. Instead you use `smtplib.SMTP` and then call `server.starttls`.

Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: Were you looking at this line: " where SSL is required from the beginning of the connection and using starttls() is not appropriate." I'm interpreting this as only use `smtplib.STMP_SSL` when TLS is not required.

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: Set TLS to default

File: tamr_toolbox/notifications/emails.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: This is a great. I added these parameters as optional

Comment (along528) [reply tamr_toolbox/notifications/emails.py]: this will also need to be passed to `_send_job_status_message` and `monitor_job`

File: tamr_toolbox/notifications/emails.py
Comment (along528) [reply tamr_toolbox/notifications/emails.py]: ah, sorry misread that then

File: tamr_toolbox/notifications/emails.py
Comment (along528) [reply tamr_toolbox/notifications/emails.py]: Also to `SMTP_SSL`, not just `starttls`

Comment (tristan-koopman) [reply tamr_toolbox/notifications/emails.py]: Thanks. I updated the last commit with these changes

**Review by along528 — APPROVED** [review]
LGTM!

Please also wait for @skalish to approve before merging since he has also provided comments

File: tamr_toolbox/notifications/emails.py
Comment (skalish) [inline tamr_toolbox/notifications/emails.py]: Can you improve this type hint to be more specific about the return?  For example, a dictionary with string-only keys and lists of integers as values would be `Dict[str, List[int]]`.

File: tamr_toolbox/notifications/emails.py
Comment (skalish) [inline tamr_toolbox/notifications/emails.py]: `Raises` is intended to report what kind of exceptions might be raised when using this function.  For example, it looks like an `SMTPException` can be expected in some cases.

What you have here looks like a return of some sort.

File: tamr_toolbox/notifications/emails.py
Comment (skalish) [inline tamr_toolbox/notifications/emails.py]: Ditto on more specific type-hinting.

File: tamr_toolbox/notifications/emails.py
Comment (skalish) [inline tamr_toolbox/notifications/emails.py]: Ditto on more specific type-hinting.

File: tamr_toolbox/notifications/slack.py
Comment (skalish) [inline tamr_toolbox/notifications/slack.py]: It'd be great to also improve these existing type-hints if possible.

File: tamr_toolbox/notifications/slack.py
Comment (skalish) [inline tamr_toolbox/notifications/slack.py]: Ditto

File: tamr_toolbox/notifications/slack.py
Comment (skalish) [inline tamr_toolbox/notifications/slack.py]: Ditto

**Review by skalish — CHANGES_REQUESTED** [review]
This looks really good!  I only have comments about docstrings and type hints.  Some improvements might not be possible depending on the actual returns.

File: tamr_toolbox/utils/operation.py
Comment (skalish) [inline tamr_toolbox/utils/operation.py]: timeout parameters should have type `float` as in `notification.emails.monitor_job`
```suggestion
    operation: Operation, *, poll_interval_seconds: float = 1, timeout_seconds: float = 300,
```

File: tamr_toolbox/notifications/common.py
Comment (skalish) [inline tamr_toolbox/notifications/common.py]:
```suggestion
    send_message: Callable,
    send_status_function: Callable,
```
If these are functions, there is a specific `typing` object for that.  `Callable` defaults to `Callable[..., Any]` which means a function that takes any number of arguments of any type and returns any type of value.  Ideally, you would specify these but it may be hard in this case.

File: tamr_toolbox/notifications/common.py
Comment (skalish) [inline tamr_toolbox/notifications/common.py]: Is this function user facing?  If not, it should start with an underscore, i.e. `_monitor_job`.

**Review by skalish — CHANGES_REQUESTED** [review]

File: tamr_toolbox/utils/operation.py
Comment (tristan-koopman) [reply tamr_toolbox/utils/operation.py]: Updated

File: tamr_toolbox/notifications/common.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/common.py]: Updated

File: tamr_toolbox/notifications/common.py
Comment (tristan-koopman) [reply tamr_toolbox/notifications/common.py]: Updated

File: tamr_toolbox/notifications/emails.py
Comment (skalish) [inline tamr_toolbox/notifications/emails.py:86]: Nit: This can be a little bit more concise, but really not an issue at all
```suggestion
    response = {"message": message}
```

**Review by skalish — APPROVED** [review]
LGTM

**Comment by along528** [comment]
@tristan-koopman You've got all approvals and checks passed so feel free to merge!

**Pull request merged**
